src/ripple/overlay/impl/OverlayImpl.h

```diff
@@ -458,39 +458,56 @@ class OverlayImpl : public Overlay
     sendEndpoints();
 
 private:
+
+    struct TrafficGauges{
+        TrafficGauges (char const* name, beast::insight::Collector::ptr const& collector)
+        : bytesIn(collector->make_gauge(name,"Bytes_In"))
+        , bytesOut(collector->make_gauge(name,"Bytes_Out"))
+        , messagesIn(collector->make_gauge(name,"Messages_In"))
+        , messagesOut(collector->make_gauge(name,"Messages_Out"))
+        {   
+        }
+        beast::insight::Gauge bytesIn;
+        beast::insight::Gauge bytesOut;
+        beast::insight::Gauge messagesIn;
+        beast::insight::Gauge messagesOut;
+    };
+
+    
     struct Stats
     {
 
         template <class Handler>
         Stats (
                 Handler const& handler, 
                 beast::insight::Collector::ptr const& collector,
-                std::vector<beast::insight::Gauge>&& trafficGauges_)
+                std::vector<TrafficGauges>&& trafficGauges_)
             : peerDisconnects (collector->make_gauge("Overlay","Peer_Disconnects"))
             , trafficGauges (trafficGauges_)
             , hook (collector->make_hook (handler))
             { 
             }
             
         beast::insight::Gauge peerDisconnects;
-        std::vector<beast::insight::Gauge> trafficGauges;
+        std::vector<TrafficGauges> trafficGauges;
         beast::insight::Hook hook;
     };
     
     Stats m_stats;
+    std::mutex m_statsMutex;
 
 private:
     void collect_metrics()
     {   
-        std::lock_guard lock (mutex_);
-        std::size_t i = 0;
+        std::lock_guard lock (m_statsMutex);
+        auto counts= m_traffic.getCounts();
 
-        for (auto const& s : m_traffic.getCounts())
+        for (std::size_t i = 0; i < counts.size(); ++i)
```

Comment:
nit: we should add an assert here that `assert(counts.size() == m_stats.trafficGauges.size())` We know it is the same from the constructor, but it's just good to put those assertions into code.

Reply:
Added, thanks!